fennel/client_tests/branches/test_clone.py

```diff
@@ -0,0 +1,390 @@
+from datetime import datetime
+
+import pandas as pd
+import pytest
+
+from fennel._vendor import requests
+from fennel.datasets import Dataset, dataset, field, pipeline
+from fennel.featuresets import featureset, feature
+from fennel.lib.schema import inputs
+from fennel.lib.aggregate import Count
+from fennel.sources import source, Webhook
+from fennel.test_lib import mock
+
+wh = Webhook(name="fennel_webhook")
+__owner__ = "nitin@fennel.com"
+
+
+@source(wh.endpoint("UserInfoDataset"))
+@dataset
+class UserInfoDataset:
+    user_id: int = field(key=True)
+    name: str
+    age: int
+    gender: str
+    country_code: int
+    email: str
+    timestamp: datetime = field(timestamp=True)
+
+
+@dataset
+class GenderStats:
+    gender: str = field(key=True)
+    count: int
+    timestamp: datetime = field(timestamp=True)
+
+    @pipeline(version=1)
+    @inputs(UserInfoDataset)
+    def my_pipeline(cls, user_info: Dataset):
+        return user_info.groupby("gender").aggregate(
+            Count(window="forever", into_field="count")
+        )
+
+
+@dataset
+class CountryStats:
+    country_code: int = field(key=True)
+    count: int
+    timestamp: datetime = field(timestamp=True)
+
+    @pipeline(version=1)
+    @inputs(UserInfoDataset)
+    def my_pipeline(cls, user_info: Dataset):
+        return user_info.groupby("country_code").aggregate(
+            Count(window="forever", into_field="count")
+        )
+
+
+@featureset
+class UserInfoFeatureset:
+    user_id: str = feature(id=1)
+    name: str = feature(id=2).extract(field=UserInfoDataset.name, default="None")  # type: ignore
+    age: int = feature(id=3).extract(field=UserInfoDataset.age, default=1)  # type: ignore
+    gender: str = feature(id=4).extract(field=UserInfoDataset.gender, default="None")  # type: ignore
+    country_code: int = feature(id=5).extract(field=UserInfoDataset.country_code, default=1)  # type: ignore
+    email: str = feature(id=6).extract(field=UserInfoDataset.email, default="None")  # type: ignore
+
+
+def _get_changed_datasets(filter_condition):
+    @source(wh.endpoint("UserInfoDataset"))
+    @dataset
+    class UserInfoDataset:
+        user_id: int = field(key=True)
+        name: str
+        age: int
+        gender: str
+        country_code: int
+        email: str
+        timestamp: datetime = field(timestamp=True)
+
```

Comment:
Does this unchanged dataset need to be in the local scope?

Reply:
yup